ofrak_core/ofrak/service/new_data_service.py

```diff
@@ -0,0 +1,409 @@
+from collections import defaultdict
+from typing import Dict, Iterable, Optional, List, Union, cast, Tuple, Set
+
+from dataclasses import dataclass
+from sortedcontainers import SortedList
+
+from ofrak.model.data_model import (
+    DataModel,
+    DataPatch,
+    DataMove,
+    DataPatchesResult,
+    DataPatchResult,
+)
+from ofrak.service.data_service_i import DataServiceInterface
+from ofrak.service.error import OutOfBoundError, PatchOverlapError
+from ofrak_type.error import NotFoundError
+from ofrak_type.range import Range
+
+# Type alias; typechecker makes no distinction between this and bytes. It's just for humans (you?).
+DataId = bytes
+
+
+class NewDataService(DataServiceInterface):
+    def __init__(self):
+        self._model_store: Dict[DataId, DataModel] = dict()
+        self._roots: Dict[DataId, _DataRoot] = dict()
+
+    def _get_by_id(self, data_id: DataId) -> DataModel:
+        model = self._model_store.get(data_id)
+        if model is None:
+            raise NotFoundError(f"No data model with ID {data_id.hex()} exists")
+        else:
+            return model
+
+    def _get_root_by_id(self, data_id: DataId) -> "_DataRoot":
+        root = self._roots.get(data_id)
+        if root is None:
+            raise NotFoundError(f"No data root with ID {data_id.hex()} exists")
+        else:
+            return root
+
+    def _is_root(self, data_id: DataId) -> bool:
+        return data_id in self._roots
+
+    def _get_absolute_range(self, id_or_model: Union[DataId, DataModel], r: Range) -> Range:
+        if isinstance(id_or_model, DataId):
+            model = self._get_by_id(cast(DataId, id_or_model))
+        else:
+            model = cast(DataModel, id_or_model)
+
+        if r.start < 0 or r.end > model.range.end:
+            raise OutOfBoundError(
+                f"The requested range {r} of model {model.id.hex()} is outside the "
+                f"model's range {model.range}"
+            )
+
+        if model.root_id is None:
+            return r
+        else:
+            root = self._get_root_by_id(model.root_id)
+            absolute_range = r.translate(model.range.start)
+            if absolute_range.end > root.model.range.end:
+                raise OutOfBoundError(
+                    f"The requested range {r} of model {model.id.hex()} is outside the "
+                    f"root's range {root.model.range}"
+                )
+            return absolute_range
```

Comment:
I would opt to moving these helper functions to the bottom.

As a reader, I'd like to read through the API first. `_get_by_id` should be pretty explanatory: if I need to dig into its implementation details, I can then look at the bottom of the class.